README.md
# Dialog

---

[![Build Status](https://img.shields.io/travis/aralejs/dialog.svg?style=flat)](https://travis-ci.org/aralejs/dialog)
[![Coverage Status](https://img.shields.io/coveralls/aralejs/dialog.svg?style=flat)](https://coveralls.io/r/aralejs/dialog)

基础对话框组件，提供对话框显示隐藏、dom 结构自定义、定位、select 遮挡、确定取消关闭等功能特性。

本模块分为两个子组件：Dialog 和 ConfirmBox，前者提供抽象的基础容器功能，无绑定 UI 样式，后者在前者的基础上提供模态对话框的功能并提供完善的 UI 样式。

Dialog 组件提供了默认样式，请单独引用 [dialog.css](https://github.com/aralejs/dialog/blob/master/src/dialog.css)。

<img width="600" src="https://t.alipayobjects.com/images/T1bexeXgxiXXXXXXXX.png">

Dialog 继承了 [overlay](http://aralejs.org/overlay/)，可使用其中包括 [widget](http://aralejs.org/widget/)、[base](http://aralejs.org/base/)、[class](http://aralejs.org/class/)、[events](http://aralejs.org/events/)、[attribute](http://aralejs.org/base/docs/attribute.html)、[aspect](http://aralejs.org/base/docs/aspect.html) 的属性和方法。

---

## 组件演示

```
npm run doc
```

## 最佳实践

```js
var Dialog = require('dialog');

var o = new Dialog({
    trigger: '#trigger',
    content: '<div>这是 dialog 容器的内容</div>'
});
```

内嵌 Iframe：

```js
var Dialog = require('dialog');

var o = new Dialog({
    trigger: '#trigger',
    content: 'https://www.alipay.com/'
});
```

## 配置说明

### trigger *element*

对话框触发元素，可传递选择器。

### content *string|element*

容器的内容，可以是纯字符串、dom对象、jQuery对象、html标签字符串、以及 URL 地址。当 content 为 URL 地址时，本组件将内嵌目标页面的 Iframe。

### hasMask *boolean*

是否有背景遮罩层。默认为 `true`。

### classPrefix *string*

统一样式前缀，默认为 `ui-dialog`。

### closeTpl *string*

右上角的关闭链接，默认为`×`。

### width *number|string*

对话框宽度，默认 500px。

### height *number|string*

对话框高度。当设置了这个属性时，对话框高度会固定，
下面的 autoFit 强制为 false，initialHeight 失效。

### zIndex *number*

对话框 z-index 层级，默认 999。

### effect *string*

简单的动画效果，none 为无动画，fade 为渐入效果。默认为 none。

### autoFit *boolean*

内嵌 Iframe 页面时是否自适应高度，默认为 true。`iframe 类型时有效`

### initialHeight *number|string*

内嵌 Iframe 页面时对话框的初始高度（onload 后会发生变化），默认为 300px 。`只有当对话框是 iframe 类型时有效`


> 其他配置如定位参数 `align` 等请参照 [overlay](http://aralejs.org/overlay/)。


## 实例方法

主要有 show、hide、render 等方法，请参照 [overlay](http://aralejs.org/overlay/)。


## 事件说明

### complete:show

本事件在当 content 为 URL 地址时 Iframe 载入完毕后触发。

```js
dialogInstanse.on('complete:show', function() {
    // Iframe 载入完毕
});
```

> close 事件已经删除，可以使用 `.after('hide', function() {})` 来代替。


## 委托绑定

需要动态加载 trigger 时，可以使用如下方式进行委托绑定。

```js
// 不初始化 trigger
var d = new Dialog({
  content: '内容'
});

$('document').on('click', '.selector', function() {
  d.activeTrigger = $(this);
  d.show();
});
```



HISTORY.md
# 历史记录

---
## 2.0.0
`changed` 迁至atool
`changed` fixed依赖templable问题

## 1.6.0

修复example

## 1.5.9

`fixed` 修复 spm@3.9 spm doc执行不通过问题。

## 1.5.2

`fixed` 修复 IE8 下 iframe 模式报错 `未指明的错误` 的问题。

## 1.5.1

`fixed` #92 修复 IE 低版本的 indexOf 的兼容问题。

## 1.5.0

`new` #91 使用 arale-messenger 进行跨域支持。[演示](./examples/cross-domain-iframe.html)

`fixed` #87 #88 彻底修复两个对话框关闭时的遮罩显示问题。

## 1.4.1

`fixed` [#87](https://github.com/aralejs/dialog/pull/87) 修复两个对话框关闭时的遮罩显示问题。

`fixed` #80 修复 top 为负数的场景。

`fixed` #71 删除了多余的 arale-easing 依赖。

## 1.4.0

`changed` 迁移到 spm@3.x
`changed` 不再默认引入样式。

## 1.3.0

`new` 增加了 ajax 载入页面内容的[方式](http://aralejs.org/dialog/examples/ajax.html)。

## 1.2.6

`improved` 升级 templatable 到 0.9.2 。

`improved` 在文档中增加取消 ESC 关闭浮层的[功能说明](http://aralejs.org/dialog/examples/custom.html#6-%E5%8F%96%E6%B6%88-esc-%E5%85%B3%E9%97%AD%E6%B5%AE%E5%B1%82%E7%9A%84%E5%8A%9F%E8%83%BD)。

## 1.2.5

`improved` 升级 overlay 到 1.1.4。

`improved` 关闭按钮的 title 调整为英文。

`improved` 向上微调了对话框的出现位置。

`new` ConfirmBox 增加 cancel 事件，并且支持 ConfirmBox.confirm(msg, title, onConfirm, onCancel, options) 的调用方式。

## 1.2.4

`tag:fixed` 去掉了 JS 代码中写死的 content 的白色背景。

## 1.2.3

`tag:fixed` #47

## 1.2.2

`tag:fixed` #44 #38 修改 content 内容导致高度变化应重新定位对话框。

`tag:fixed` #43 多个对话框共享一个 mask 的后续 confirmbox 问题。

## 1.2.1

`tag:improved` 去掉了对话框的虚线框，并优化了无障碍体验。[参考](http://ued.taobao.com/blog/2011/04/onfocus-this-blur/)

`tag:fixed` #40 修复一个 iframe 对话框设置高度的问题。

`tag:fixed` 升级 overlay 到 1.1.3，修复遮罩层无法挡住 select 的问题。

`tag:fixed` 修复同一个实例多次调用 show 后无法隐藏遮罩层的问题。（这个问题在 1.2.0 时引入）

`tag:changed` #41 增加了高度超过一屏时的位置处理。


## 1.2.0

`tag:new` 新增 initialHeight 属性，用于设置 iframe 对话框的初始化高度。

`tag:changed` 移除 hasMask.hideOnClick 属性。

`tag:changed` 去掉对 mask 的 z-index 设置，改为调整 mask 的 dom 位置。

`tag:improved` #37 支持多个对话框共享同一个 mask 的[情况](examples/two-dialog.html)。

## 1.1.3

`tag:improved` overlay 升级到 1.1.2，从而修复一个 IE 冗余调用 resize 的问题。

`tag:changed` 修改 confirmbox 的 hasMask 默认值，使其点击遮罩层默认不关闭。

## 1.1.2

`tag:improved` overlay 升级到 1.1.1, mask 升级到 1.1.1, templatable 升级到 0.9.1

`tag:fixed` #33 修复 firefox 下确定取消按钮文字偏下的问题。

`tag:new` #34 #35 添加点击遮罩层时关闭对话框的特性，并改变了 hasMask 的默认值为 `{hideOnClick: true}`。

`tag:improved` 添加 styleBox, 已解决样式冲突问题

## 1.1.1

`tag:fixed` #32 修复由于本地缓存导致 position 依赖打包重复的问题。

## 1.1.0

`tag:improved` 对 arale/overlay 的依赖从 1.0.1 升级到 1.1.0 。

`tag:improved` 对 arale/mask 的依赖从 1.0.1 升级到 1.1.0 。

`tag:improved` 对 arale/events 的依赖从 1.0.0 升级到 1.1.0 。

`tag:improved` 对 templatable 的依赖从 arale/widget/1.0.3/templatable 升级到 arale/templatable/0.9.0/templatable 。

`tag:improved` 对 handlebars 的依赖从 gallery/handlebars/1.0.0/handlebars 升级到 gallery/handlebars/1.0.2/runtime 。

## 1.0.3

`tag:fixed` 修复 ConfirmBox.show 的回调失效问题。

## 1.0.2

`tag:fixed` [#25](https://github.com/aralejs/dialog/issues/25) 使得 dialog 对 mask 的 z-index 的处理不影响其他实例。

`tag:fixed` [pull/24](pull/24) 修复 ConfirmBox 静态方法生成的对象销毁失效的问题。

`tag:fixed` 修复 ConfirmBox 由于 1.0.1 版本的 dist 文件打包错误导致无法使用的问题。

## 1.0.1

`tag:fixed` [#21](https://github.com/aralejs/dialog/issues/21) 修复 IE6 下内容区域的高度撑不开的问题。

`tag:fixed` [#20](https://github.com/aralejs/dialog/issues/20) 修复 IE6 下换不同 url 时会产生 404 页面的问题。

`tag:changed` 去掉 close 自定义事件，可以用 before/after('hide') 来代替。

`tag:improved` 对 arale/overlay 的依赖升级到 1.0.1。

`tag:unresolved` 去掉 close 自定义事件导致 ConfirmBox.alert 等静态方法产生的对话框无法销毁自身。

## 1.0.0

本版本对 dialog 进行了毁灭性的重构，API 和定位都发生了大量变化，基本和之前的版本没什么关系。若要查看旧版本的文档，请到仓库中 clone 对应的 tag 到本地进行查看。

`tag:changed` BaseDialog 改名为 Dialog，定位为对话框容器，支持 Iframe 内嵌，自带模板以及样式内联。

`tag:changed` 去掉 AnimDialog，改为只支持简单的渐入效果。

`tag:changed` ConfirmBox API 大幅重构，定位为标准模态对话框。

`tag:improved` 对 arale/overlay 的依赖从 0.9.13 升级到 1.0.0

`tag:improved` 对 arale/widget 的依赖从 1.0.2 升级到 1.0.3

`tag:improved` 自动计算 mask 遮罩层的 z-index 。

`tag:fixed` 修复页面跳转会停止自适应高度的问题

## 0.9.3

`tag:improved` AnimDialog 增加 confirm 和 close 的事件阻止，防止打断可能出现的异步请求。

## 0.9.2

`tag:changed` [#10](https://github.com/aralejs/dialog/issues/10) ConfirmBox.confirm 等静态方法的参数调整，支持传自定义参数。

`tag:improved` ConfirmBox 隐藏后销毁自身。

`tag:improved` overlay 依赖升级到 0.9.13。

## 0.9.1

`tag:changed` 删除 confirmBox.message 方法。

`tag:changed` [#9](https://github.com/aralejs/dialog/issues/9) 去掉回车确定的特性。

`tag:improved` content  和 title 支持传动态 function。

`tag:improved` [#7](https://github.com/aralejs/dialog/issues/7) 去掉titleElement中对h2标签的依赖。

`tag:new` [#6](https://github.com/aralejs/dialog/issues/6) 新增confirm和close两个方法。

## 0.9.0

`tag:new` 新组件。


src/dialog.js
import './dialog.css';
import $ from 'jquery';
import Overlay from 'arale-overlay';
import Events from 'arale-events';
import Templatable from 'arale-templatable';
import Messenger from 'arale-messenger';

const mask = Overlay.Mask;
// Dialog
// ---
// Dialog 是通用对话框组件，提供显隐关闭、遮罩层、内嵌iframe、内容区域自定义功能。
// 是所有对话框类型组件的基类。
const Dialog = Overlay.extend({

  Implements: Templatable,

  attrs: {
    // 模板
    template: `<div class="{{classPrefix}}">
                  <a class="{{classPrefix}}-close" title="Close" href="javascript:;" data-role="close"></a>
                  <div class="{{classPrefix}}-content" data-role="content"></div>
              </div>`,

    // 对话框触发点
    trigger: {
      value: null,
      getter: function (val) {
        return $(val);
      }
    },

    // 统一样式前缀
    classPrefix: 'ui-dialog',

    // 指定内容元素，可以是 url 地址
    content: {
      value: null,
      setter: function (val) {
        // 判断是否是 url 地址
        if (/^(https?:\/\/|\/|\.\/|\.\.\/)/.test(val)) {
          this._type = 'iframe';
          // 用 ajax 的方式而不是 iframe 进行载入
          if (val.indexOf('?ajax') > 0 || val.indexOf('&ajax') > 0) {
            this._ajax = true;
          }
        }
        return val;
      }
    },

    // 是否有背景遮罩层
    hasMask: true,

    // 关闭按钮可以自定义
    closeTpl: '×',

    // 默认宽度
    width: 500,

    // 默认高度
    height: null,

    // iframe 类型时，dialog 的最初高度
    initialHeight: 300,

    // 简单的动画效果 none | fade
    effect: 'none',

    // 不用解释了吧
    zIndex: 999,

    // 是否自适应高度
    autoFit: true,

    // 默认定位左右居中，略微靠上
    align: {
      value: {
        selfXY: ['50%', '50%'],
        baseXY: ['50%', '42%']
      },
      getter: function (val) {
        // 高度超过窗口的 42/50 浮层头部顶住窗口
        // https://github.com/aralejs/dialog/issues/41
        if (this.element.height() > $(window).height() * 0.84) {
          return {
            selfXY: ['50%', '0'],
            baseXY: ['50%', '0']
          };
        }
        return val;
      }
    }
  },


  parseElement: function () {
    this.set("model", {
      classPrefix: this.get('classPrefix')
    });
    Dialog.superclass.parseElement.call(this);
    this.contentElement = this.$('[data-role=content]');

    // 必要的样式
    this.contentElement.css({
      height: '100%',
      zoom: 1
    });
    // 关闭按钮先隐藏
    // 后面当 onRenderCloseTpl 时，如果 closeTpl 不为空，会显示出来
    // 这样写是为了回避 arale.base 的一个问题：
    // 当属性初始值为''时，不会进入 onRender 方法
    // https://github.com/aralejs/base/issues/7
    this.$('>[data-role=close]').hide();
  },

  events: {
    'click [data-role=close]': function (e) {
      e.preventDefault();
      this.hide();
    }
  },

  show: function () {
    // iframe 要在载入完成才显示
    if (this._type === 'iframe') {
      // ajax 读入内容并 append 到容器中
      if (this._ajax) {
        this._ajaxHtml();
      } else {
        // iframe 还未请求完，先设置一个固定高度
        !this.get('height') && this.contentElement.css('height', this.get('initialHeight'));
        this._showIframe();
      }
    }

    Dialog.superclass.show.call(this);
    return this;
  },

  hide: function () {
    // 把 iframe 状态复原
    if (this._type === 'iframe' && this.iframe) {
      // 如果是跨域iframe，会抛出异常，所以需要加一层判断
      if (!this._isCrossDomainIframe) {
        this.iframe.attr({
          src: 'javascript:\'\';'
        });
      }
      // 原来只是将 iframe 的状态复原
      // 但是发现在 IE6 下，将 src 置为 javascript:''; 会出现 404 页面
      this.iframe.remove();
      this.iframe = null;
    }

    Dialog.superclass.hide.call(this);
    clearInterval(this._interval);
    delete this._interval;
    return this;
  },

  destroy: function () {
    this.element.remove();
    this._hideMask();
    clearInterval(this._interval);
    return Dialog.superclass.destroy.call(this);
  },

  setup: function () {
    Dialog.superclass.setup.call(this);

    this._setupTrigger();
    this._setupMask();
    this._setupKeyEvents();
    this._setupFocus();
    toTabed(this.element);
    toTabed(this.get('trigger'));

    // 默认当前触发器
    this.activeTrigger = this.get('trigger').eq(0);
  },

  // onRender
  // ---
  _onRenderContent: function (val) {
    if (this._type !== 'iframe') {
      var value;
      // 有些情况会报错
      try {
        value = $(val);
      } catch (e) {
        value = [];
      }
      if (value[0]) {
        this.contentElement.empty().append(value);
      } else {
        this.contentElement.empty().html(val);
      }
      // #38 #44
      this._setPosition();
    }
  },

  _onRenderCloseTpl: function (val) {
    if (val === '') {
      this.$('>[data-role=close]').html(val).hide();
    } else {
      this.$('>[data-role=close]').html(val).show();
    }
  },

  // 覆盖 overlay，提供动画
  _onRenderVisible: function (val) {
    if (val) {
      if (this.get('effect') === 'fade') {
        // 固定 300 的动画时长，暂不可定制
        this.element.fadeIn(300);
      } else {
        this.element.show();
      }
    } else {
      this.element.hide();
    }
  },

  // 私有方法
  // ---
  // 绑定触发对话框出现的事件
  _setupTrigger: function () {
    this.delegateEvents(this.get('trigger'), 'click', function (e) {
      e.preventDefault();
      // 标识当前点击的元素
      this.activeTrigger = $(e.currentTarget);
      this.show();
    });
  },

  // 绑定遮罩层事件
  _setupMask: function () {
    var that = this;

    // 存放 mask 对应的对话框
    mask._dialogs = mask._dialogs || [];

    this.after('show', function () {
      if (!this.get('hasMask')) {
        return;
      }
      // not using the z-index
      // because multiable dialogs may share same mask
      mask.set('zIndex', that.get('zIndex')).show();
      mask.element.insertBefore(that.element);

      // 避免重复存放
      var existed;
      for (var i=0; i<mask._dialogs.length; i++) {
        if (mask._dialogs[i] === that) {
          existed = mask._dialogs[i];
        }
      }
      if (existed) {
        // 把已存在的对话框提到最后一个
        erase(existed, mask._dialogs);
        mask._dialogs.push(existed);
      } else {
        // 存放新的对话框
        mask._dialogs.push(that);
      }
    });

    this.after('hide', this._hideMask);
  },

  // 隐藏 mask
  _hideMask: function () {
    if (!this.get('hasMask')) {
      return;
    }

    // 移除 mask._dialogs 当前实例对应的 dialog
    var dialogLength = mask._dialogs ? mask._dialogs.length : 0;
    for (var i=0; i<dialogLength; i++) {
      if (mask._dialogs[i] === this) {
        erase(this, mask._dialogs);

        // 如果 _dialogs 为空了，表示没有打开的 dialog 了
        // 则隐藏 mask
        if (mask._dialogs.length === 0) {
          mask.hide();
        }
        // 如果移除的是最后一个打开的 dialog
        // 则相应向下移动 mask
        else if (i === dialogLength - 1) {
          var last = mask._dialogs[mask._dialogs.length - 1];
          mask.set('zIndex', last.get('zIndex'));
          mask.element.insertBefore(last.element);
        }
      }
    }
  },

  // 绑定元素聚焦状态
  _setupFocus: function () {
    this.after('show', function () {
      this.element.focus();
    });
    this.after('hide', function () {
      // 关于网页中浮层消失后的焦点处理
      // http://www.qt06.com/post/280/
      this.activeTrigger && this.activeTrigger.focus();
    });
  },

  // 绑定键盘事件，ESC关闭窗口
  _setupKeyEvents: function () {
    this.delegateEvents($(document), 'keyup.esc', function (e) {
      if (e.keyCode === 27) {
        this.get('visible') && this.hide();
      }
    });
  },

  _showIframe: function () {
    var that = this;
    // 若未创建则新建一个
    if (!this.iframe) {
      this._createIframe();
    }

    // 开始请求 iframe
    this.iframe.attr({
      src: this._fixUrl(),
      name: 'dialog-iframe' + new Date().getTime()
    });

    // 因为在 IE 下 onload 无法触发
    // http://my.oschina.net/liangrockman/blog/24015
    // 所以使用 jquery 的 one 函数来代替 onload
    // one 比 on 好，因为它只执行一次，并在执行后自动销毁
    this.iframe.one('load', function () {
      // 如果 dialog 已经隐藏了，就不需要触发 onload
      if (!that.get('visible')) {
        return;
      }

      // 是否跨域的判断需要放入iframe load之后
      that._isCrossDomainIframe = isCrossDomainIframe(that.iframe);

      if (!that._isCrossDomainIframe) {
        // 绑定自动处理高度的事件
        if (that.get('autoFit')) {
          clearInterval(that._interval);
          that._interval = setInterval(function () {
            that._syncHeight();
          }, 300);
        }
        that._syncHeight();
      }

      that._setPosition();
      that.trigger('complete:show');
    });
  },

  _fixUrl: function () {
    var s = this.get('content').match(/([^?#]*)(\?[^#]*)?(#.*)?/);
    s.shift();
    s[1] = ((s[1] && s[1] !== '?') ? (s[1] + '&') : '?') + 't=' + new Date().getTime();
    return s.join('');
  },

  _createIframe: function () {
    var that = this;

    this.iframe = $('<iframe>', {
      src: 'javascript:\'\';',
      scrolling: 'no',
      frameborder: 'no',
      allowTransparency: 'true',
      css: {
        border: 'none',
        width: '100%',
        display: 'block',
        height: '100%',
        overflow: 'hidden'
      }
    }).appendTo(this.contentElement);

    // 给 iframe 绑一个 close 事件
    // iframe 内部可通过 window.frameElement.trigger('close') 关闭
    Events.mixTo(this.iframe[0]);
    this.iframe[0].on('close', function () {
      that.hide();
    });

    // 跨域则使用arale-messenger进行通信
    var m = new Messenger('parent', 'arale-dialog');
    this.iframe.one('load', function () {
      m.addTarget(that.iframe[0].contentWindow, 'iframe1');
      m.listen(function (data) {
        data = JSON.parse(data);
        switch (data.event) {
          case 'close':
            that.hide();
            break;
          case 'syncHeight':
            that._setHeight(data.height.toString().slice(-2) === 'px' ? data.height : data.height + 'px');
            break;
          default:
            break;
        }
      });
    });
  },

  _setHeight: function (h) {
    this.contentElement.css('height', h);
    // force to reflow in ie6
    // http://44ux.com/blog/2011/08/24/ie67-reflow-bug/
    this.element[0].className = this.element[0].className;
  },

  _syncHeight: function () {
    var h;
    // 如果未传 height，才会自动获取
    if (!this.get('height')) {
      try {
        this._errCount = 0;
        h = getIframeHeight(this.iframe) + 'px';
      } catch (err) {
        // 页面跳转也会抛错，最多失败6次
        this._errCount = (this._errCount || 0) + 1;
        if (this._errCount >= 6) {
          // 获取失败则给默认高度 300px
          // 跨域会抛错进入这个流程
          h = this.get('initialHeight');
          clearInterval(this._interval);
          delete this._interval;
        }
      }
      this._setHeight(h);

    } else {
      clearInterval(this._interval);
      delete this._interval;
    }
  },

  _ajaxHtml: function () {
    var that = this;
    this.contentElement.css('height', this.get('initialHeight'));
    this.contentElement.load(this.get('content'), function () {
      that._setPosition();
      that.contentElement.css('height', '');
      that.trigger('complete:show');
    });
  }

});

module.exports = Dialog;

// Helpers
// ----
// 让目标节点可以被 Tab
function toTabed(element) {
  if (element.attr('tabindex') == null) {
    element.attr('tabindex', '-1');
  }
}

// 获取 iframe 内部的高度
function getIframeHeight(iframe) {
  var D = iframe[0].contentWindow.document;
  if (D.body.scrollHeight && D.documentElement.scrollHeight) {
    return Math.min(D.body.scrollHeight, D.documentElement.scrollHeight);
  } else if (D.documentElement.scrollHeight) {
    return D.documentElement.scrollHeight;
  } else if (D.body.scrollHeight) {
    return D.body.scrollHeight;
  }
}


// iframe 是否和当前页面跨域
function isCrossDomainIframe(iframe) {
  var isCrossDomain = false;
  try {
    iframe[0].contentWindow.document;
  } catch (e) {
    isCrossDomain = true;
  }
  return isCrossDomain;
}

// erase item from array
function erase(item, array) {
  var index = -1;
  for (var i=0; i<array.length; i++) {
    if (array[i] === item) {
      index = i;
      break;
    }
  }
  if (index !== -1) {
    array.splice(index, 1);
  }
}
